stab_controller_multi_lqr_node.cpp

```diff
@@ -0,0 +1,416 @@
+    // Start rated loop
+    RateController();
+
+  }
+  StabControllerMultiLqrNode::~StabControllerMultiLqrNode() {
+    maxon_left.MoveWithTorque(0);
+    maxon_right.MoveWithTorque(0);
+  }
+
+  //Setup methods
+  void StabControllerMultiLqrNode::ResetStates() {
+    encoder_offset_left = maxon_left.GetPosition()*2*M_PI/encoder_quadcounts;
+    encoder_offset_right = maxon_right.GetPosition()*2*M_PI/encoder_quadcounts;
+  }
+  void StabControllerMultiLqrNode::Initialise() {
+    // Read parameters from server
+    if (!ReadParameters()) {
+      ROS_ERROR("Could not read parameters.");
+      ros::requestShutdown();
+    }
+
+    InitSubPub();
+    InitSrv();
+
+    // Setup Controller
+    ResetStates();
+    Eigen::VectorXd u_sat = Eigen::MatrixXd(NO_OF_INPUTS,1);
+    u_sat << saturation_ec90;
+    controller.SetUSat(u_sat);
+    input_delays_left.setZero(NO_OF_DELAYS, 1);
+    input_delays_right.setZero(NO_OF_DELAYS, 1);
+    input_left.setZero(NO_OF_INPUTS, 1);
+    input_right.setZero(NO_OF_INPUTS, 1);
+    meas_left.setZero(NO_OF_MEASUREMENTS, 1);
+    meas_right.setZero(NO_OF_MEASUREMENTS, 1);
+
+
+    // Read in K matrices
+    ReadInMatrixArray("/home/ascento/catkin_ws/src/ascento/Matlab/DynMod/fixed_drive/ros_compatible/k_matrix_");
+    // Read in Imu offsets
+    ReadInDoubleArray("/home/ascento/catkin_ws/src/ascento/Matlab/DynMod/fixed_drive/ros_compatible/imu_offset");
+  }
+  void StabControllerMultiLqrNode::InitSubPub() {
+    //Initialise subscribers
+    imu_sub = nh.subscribe("/est_states/imu", 1, &StabControllerMultiLqrNode::SaveImuCallback, this);
+    anydrive_left_sub = nh.subscribe("/sens_states/hip_mot/left", 1, &StabControllerMultiLqrNode::SaveAnydriveLeftCallback, this);
+    anydrive_right_sub = nh.subscribe("/sens_states/hip_mot/right", 1, &StabControllerMultiLqrNode::SaveAnydriveRightCallback, this);
+
+    maxon_execution_pub = nh.advertise<ascento_msgs::wheel_input_torque>("/actuate/wheel_mot", 1);
+  }
+  void StabControllerMultiLqrNode::InitSrv() {
+    //Initialise services
+    controller_starter = nh.advertiseService("/stab_control_start", &StabControllerMultiLqrNode::StartStopCallback, this);
+    imu_autotune = nh.advertiseService("/set_imu_offset", &StabControllerMultiLqrNode::SetImuOffsetCallback, this);
+  }
+  bool StabControllerMultiLqrNode::ReadParameters() {
+    // Geometric parameters
+    if (!nh.getParam("/wheel/radius", wheel_radius)) return false;
+    if (!nh.getParam("/phi/min", phi_min)) return false;
+    if (!nh.getParam("/phi/max", phi_max)) return false;
+    if (!nh.getParam("/maxon_ec90_max_torque", saturation_ec90)) return false;
+    if (!nh.getParam("/encoder/quadcounts", encoder_quadcounts)) return false;
+    if (!nh.getParam("/measured_imu_offset", measured_imu_offset)) ROS_INFO("No param \"/measured_imu_offset\" found, run srv \"/set_imu_offset\"");
+
+    // Controller Parameters
+    if (!nh.getParam("/stab_controller_multi_lqr/sampling_rate", sampling_rate)) return false;
```

Comment:
don't hard code absolute topic paths like this nh.getParam("sampling_rate" ... will work and has the advantage of still working if the node is in a group or has its name changed.